### PR TITLE
[Inductor] Fix stale backed-symbol references in AOTI deferred runtime asserts (#184624)

### PR DESCRIPTION
**Pull request by rmiao**
Summary:

## Problem

When multiple inputs share the same dynamic batch dimension (via `dynamic_shapes`) and are sliced with the same `nonzero().shape[0]` result, AOTI compilation fails with C++ errors like `use of undeclared identifier 's77'`.

## Root Cause

During `torch.export`, ShapeEnv creates a separate backed symbol for each input's batch dimension (e.g., `s77` for `x.size()[0]`, `s97` for `a.size()[0]`). The constraint solver then discovers they're all equal (shared dim `B`) and records replacements: `s77 -> s31`, `s97 -> s31`, etc.

After constraint solving, `codegen_input_symbol_assignment` uses `i.node.expr` (which applies `shape_env.replace()`) and defines only the canonical symbol `s31`. However, `deferred_runtime_asserts` were created before constraint solving and still reference pre-replacement symbols like `s77`. When these are emitted as `AssertScalar` IR nodes, the generated C++ references undefined variables.

## Fix

Three changes:

1. `graph.py`: Rename `symbolic_sizes_strides` to `get_placeholder_sizes_strides` and use `_get_placeholder_expr(i.node)` to declare input sizes/strides with the raw pre-replacement symbol (e.g., `s77` from `._expr`) instead of the post-replacement canonical (e.g., `s31` from `.expr`). This ensures codegen defines all symbols that deferred runtime asserts reference. A backward-mode guard preserves the existing `.expr` behavior as a workaround for https://github.com/pytorch/pytorch/issues/155468.

2. `meta_utils.py`: Use `guard_or_false(sym_eq(...))` for stride/offset comparison in `from_tensor()` to avoid `GuardOnDataDependentSymNode` crash when FakeTensors have unbacked SymInts from `nonzero()`.

3. `test_aot_inductor.py`: New regression test `test_multi_input_nonzero_slice_shared_dim` covering the shared-dynamic-dim + nonzero scenario.

Test Plan:
All tests pass locally on devgpu (A100):

**GPU ABI-compatible (new regression test):**
```
buck2 run fbcode//mode/opt -c 'fbcode.nvcc_arch=a100,h100' fbcode//caffe2/test/inductor:test_aot_inductor -- -r test_multi_input_nonzero_slice_shared_dim
```
Ran 3, OK (skipped=1).

**CPU arrayref (stack_allocation + minimal_arrayref_interface):**
```
buck2 run fbcode//mode/opt fbcode//caffe2/test/inductor:aot_inductor_arrayref_cpu -- -r test_multi_input_nonzero_slice_shared_dim
```
Ran 2, OK.

**Existing unbacked symint tests (GPU+CPU):**
```
buck2 run fbcode//mode/opt -c 'fbcode.nvcc_arch=a100,h100' fbcode//caffe2/test/inductor:test_aot_inductor -- -r test_unbacked
```
Ran 30, OK (skipped=18).

**Runtime assert tests (verifies unbacked symbol assertions still fire):**
```
buck2 run fbcode//mode/opt fbcode//caffe2/test:dynamic_shapes -- -r test_post_specialize_runtime_assert1
```
Ran 2, OK.

Reviewed By: kqfu, desertfire

Differential Revision: D105336317

cc @voznesenskym @penguinwu @EikanWang @jgong5 @Guobing-Chen @XiaobingSuper @zhuhaozhe @blzheng @wenzhe-nrv @jiayisunx @ipiszy @kadeng @muchulee8 @amjames @chauhang @aakhundov @coconutruben @jataylo

### CODE REVIEW

**Review by kqfu — APPROVED** [review]
Looks good

File: test/inductor/test_aot_inductor.py [hunk 1]
```diff
@@ -8911,6 +8945,8 @@ def fail_gpu(suffixes: tuple[str, ...], is_skip=False):
     "test_quantized_linear": fail_gpu(("cuda", "xpu")),
     "test_quanatized_int8_linear": fail_gpu(("cuda", "xpu")),
     "test_quantized_linear_bias_none": fail_gpu(("cuda", "xpu")),
+    # pattern_matcher _convert_to_fake_mode guards on data-dependent nonzero
+    "test_multi_input_nonzero_slice_shared_dim": fail_gpu(("cuda",), is_skip=True),
```
Comment (desertfire) [inline test/inductor/test_aot_inductor.py]: Why do we skip both CPU and GPU here?

File: torch/_inductor/graph.py [hunk 2]
```diff
@@ -2301,7 +2301,15 @@ def is_convertible(s: Expr) -> bool:
                         i1 = min(missing, key=str)
                         self.ras_by_symbol.setdefault(i1, []).append(ra)
                     else:
-                        make_assert(ra.expr, f"{ra.expr}")
+                        expr = ra.expr
+                        if not fvs:
+                            # Pure backed-symbol assertion: substitute
+                            # replacements so codegen references defined
+                            # symbols. Skip tautologies (e.g., Eq(s31, s31)).
+                            expr = expr.xreplace(shape_env.replacements)
```
Comment (desertfire) [inline torch/_inductor/graph.py]: Better to use existing util function, `shape_env.replace`.

**Review by desertfire — DISMISSED** [review]
Please fix lint errors and CI errors (could be adding more skippings)

File: test/inductor/test_aot_inductor.py (hunk 1 above)
Comment (rmiao) [reply test/inductor/test_aot_inductor.py]: Good point! Fixed by catching GuardOnDataDependentSymNode in pattern_matcher.py and return None to skip the pattern match not the test itself.

File: torch/_inductor/graph.py (hunk 2 above)
Comment (rmiao) [reply torch/_inductor/graph.py]: shape_env.replace() cannot be used here because it replaces all symbols including unbacked ones, which turns data-dependent runtime assertions into tautologies (e.g., u0 == s0+s1 becomes (s0+s1) == s0+s1 -> True, silently removing the check). We use _find(), which is the same DSU utility that replace() uses internally. But it is only for backed symbols.

File: torch/_inductor/pattern_matcher.py [hunk 3]
```diff
@@ -369,7 +369,10 @@ def record(node: torch.fx.Node, val: Any) -> None:
 
                     def _convert_to_fake_mode(it):
                         if isinstance(it, FakeTensor) and fake_mode is not None:
-                            return fake_mode.from_tensor(it)
+                            try:
+                                return fake_mode.from_tensor(it)
+                            except torch.fx.experimental.symbolic_shapes.GuardOnDataDependentSymNode:
```
Comment (desertfire) [inline torch/_inductor/pattern_matcher.py]: Can you help me to understand why we need to catch this exception now?

File: torch/_inductor/graph.py [hunk 4]
```diff
@@ -2301,7 +2302,24 @@ def is_convertible(s: Expr) -> bool:
                         i1 = min(missing, key=str)
                         self.ras_by_symbol.setdefault(i1, []).append(ra)
                     else:
-                        make_assert(ra.expr, f"{ra.expr}")
+                        from torch.utils._sympy.symbol import symbol_is_type, SymT
+
+                        repls = {}
```
Comment (desertfire) [inline torch/_inductor/graph.py]: Nit: let's make the following as a util function.

File: torch/_inductor/pattern_matcher.py (hunk 3 above)
Comment (rmiao) [reply torch/_inductor/pattern_matcher.py]: added a comment in the code. basically, FakeTensors with unbacked SymInts can't be converted between fake modes.

Comment (laithsakka) [reply torch/_inductor/pattern_matcher.py]: we shall have handling for this it shall work let me take deeper look. does the unit test hit the issue here

File: torch/_inductor/graph.py [hunk 5]
```diff
@@ -2301,7 +2302,10 @@ def is_convertible(s: Expr) -> bool:
                         i1 = min(missing, key=str)
                         self.ras_by_symbol.setdefault(i1, []).append(ra)
                     else:
-                        make_assert(ra.expr, f"{ra.expr}")
+                        expr = shape_env.replace_backed_only(ra.expr)
```
Comment (laithsakka) [inline torch/_inductor/graph.py]: the right fix is that input dependent code gen  shall use ._expr instead with no replacements

**Review by laithsakka — DISMISSED** [review]

File: torch/_inductor/pattern_matcher.py (hunk 3 above)
Comment (laithsakka) [reply torch/_inductor/pattern_matcher.py]: for this part this is what you need

Image

File: torch/_inductor/graph.py (hunk 5 above)
Comment (rmiao) [reply torch/_inductor/graph.py]: Thanks for the suggestion. ra.expr here is a plain SympyBoolean, not a SymNode, so ._expr doesn't directly apply. 
  
The root issue is that these deferred runtime asserts are created before constraint solving discovers equalities. After solving, codegen_input_symbol_assignment only defines the canonical symbol, but ra.expr still references the stale pre-replacement symbol.

The new util function replace_backed_only (as suggested by previous comment) uses _find() (the same DSU utility that replace() uses internally) to canonicalize only the backed symbols, preserving unbacked symbols so data-dependent runtime assertions still fire. Happy to iterate if you have a specific alternative approach in mind.

Comment (laithsakka) [reply torch/_inductor/graph.py]: something like this
https://github.com/pytorch/pytorch/pull/185570

Comment (rmiao) [reply torch/_inductor/graph.py]: Thanks for the comment. I have addressed the issues and PTAL. Thanks a lot!
